src/main/java/htsjdk/samtools/cram/CRAIIndex.java

```diff
@@ -128,16 +86,16 @@ public static SeekableStream openCraiFileAsBaiStream(final InputStream indexStre
 
         for (final CRAIEntry entry : full) {
             final Slice slice = new Slice();
-            slice.containerOffset = entry.containerStartOffset;
-            slice.alignmentStart = entry.alignmentStart;
-            slice.alignmentSpan = entry.alignmentSpan;
-            slice.sequenceId = entry.sequenceId;
-            // https://github.com/samtools/htsjdk/issues/531
-            // entry.sliceSize is the slice size in bytes, not the number of
-            // records; this results in the BAMIndex metadata being wrong
-            slice.nofRecords = entry.sliceSize;
-            slice.index = entry.sliceIndex;
-            slice.offset = entry.sliceOffset;
+            slice.containerOffset = entry.getContainerStartByteOffset();
+            slice.alignmentStart = entry.getAlignmentStart();
+            slice.alignmentSpan = entry.getAlignmentSpan();
+            slice.sequenceId = entry.getSequenceId();
+            // NOTE: the recordCount and sliceIndex fields can't be derived from the CRAM index
+            // so we can only set them to zero
+            // see https://github.com/samtools/htsjdk/issues/531
+            slice.nofRecords = 0;
+            slice.index = 0;
+            slice.offset = entry.getSliceByteOffset();
 
             indexer.processSingleReferenceSlice(slice);
```

Comment:
I was just starting to come around to your thinking about IndexableSlice, but the more I look at this, I wonder if this indexing code should be working right off of the CRAIEntry, instead of creating a partial slice, and then using the exact subset of slice that was populated from the CRAIEntry ? I wonder if we can cut out the slice altogether ?

Reply:
Just to clarify, in this case its converting to BAI, so there may be no need to use a slice for that. There are probably still other cases where we want to create an index from a real slice, in which case we'd use the real thing.

Reply:
In my Slice PR I have `SliceBAIMetadata` playing this role: essentially just `CRAIEntry` with the necessary addition of `nofRecords` and `index`.